### DISCUSSION TITLE
Support Postgres Interval in GraphQL

### DISCUSSION THREAD

**Discussion by Joey92** [directus/directus#12841] [Feature Requests]
### Preflight Checklist

- [X] I have completed all [Troubleshooting Steps](https://docs.directus.io/getting-started/support/#troubleshooting-steps).
- [X] I'm on [the latest version of Directus](https://github.com/directus/directus/releases).
- [X] There's [no other issue](https://github.com/directus/directus/issues?q=is%3Aissue) that already describes my problem.

### Describe the Bug

When I want to access fields via GraphQL that are of type interval in Postgres the values are null.
In the GraphQL errors I get a message like this: `String cannot represent value: { hours: 22, minutes: 36, seconds: 55, milliseconds: 220 }` 
When querying it as a REST call it works fine.

I guess the pg client for node gives you the data as an object, but GraphQL takes the type of the field configured in Directus, which seems to be treating it as a string. (My educated guess here)

### To Reproduce

Create a field in Postgres of type Interval, not via Directus admin UI. Add some data and query the collection via GraphQL.

### Errors Shown

Browser console: String cannot represent value: { hours: 22, minutes: 36, seconds: 55, milliseconds: 220 }

### What version of Directus are you using?

9.8.0

### What version of Node.js are you using?

16.9.1

### What database are you using?

Postgres 13.4

### What browser are you using?

Firefox

### What operating system are you using?

Linux

### How are you deploying Directus?

As project dependency in package.json, runtime in node16-alpine docker container

**Comment by azrikahar** [comment]
> I guess the pg client for node gives you the data as an object, but GraphQL takes the type of the field configured in Directus, which seems to be treating it as a string. (My educated guess here)

You're right on the mark here! Currently `interval` is treated/"mapped" as `string` by default:

https://github.com/directus/directus/blob/703fb8426f59619a4f21e80f670ed4347d01666c/api/src/utils/get-local-type.ts#L84

and then for GraphQL, a util function maps the field type to the appropriate GraphQL type, but in this case, it defaults to `GraphQLString`:

https://github.com/directus/directus/blob/703fb8426f59619a4f21e80f670ed4347d01666c/api/src/utils/get-graphql-type.ts#L14-L37

---

Changing the 1st referenced code to `json` does resolve your issue, but I'm wondering how can we make the experience better on the app side, since treating it as json ended up showing the code interface:

![image](https://user-images.githubusercontent.com/42867097/161488610-360b93fb-812c-4ebb-b552-97d9b2cdd437.png)

Although on surface it seems like we can edit it just fine like a JSON, but it is in fact not and should not be treated as such (or else we'll have weird results), as stated by this SO answer: https://stackoverflow.com/a/58876854. Here's the "weirdness" in action, where hour gets saved as minutes instead:

https://user-images.githubusercontent.com/42867097/161489300-699fe9f4-a564-4071-904e-51571142c1fa.mp4

Saving it as a string like "2h 30m" despite it is a code interface does work! However there's also currently the issue #11098 that is disrupting this workaround at the moment.

---

@Joey92 this isn't ideal and more so a temporary workaround, but can you try going to `directus_fields` and update the `special` column for your interval field row to `cast-json`? It should make it work for both REST and GraphQL, just that the value editing experience on the App side is not perfect.

**Comment by Joey92** [comment]
Wow thank you for your detailed response!

Treating the field as a string is actually not that bad. It's an object because of the node pg client. In the database itself you can also insert into an interval field as a string `30:00`. There are a bunch of ways the database could be fed by casting a string to interval on query level: https://www.postgresql.org/docs/current/datatype-datetime.html#DATATYPE-INTERVAL-INPUT-EXAMPLES

I'll try out your suggestion setting the field to `cast-json` later today. It's a read-only field in my app, so no need to worry about the editing experience. :)

**Comment by rijkvanzanten** [comment]
This is _technically_ more of a feature request to properly support Interval types in the platform, but lets first treat it as a bug to make sure it's fallback support is returned consistently across both rest and graphql 🙂

**Comment by Joey92** [comment]
@rijkvanzanten Yep I was really unsure on where to put it. I choose Bug since it's technically working with REST but not with GraphQL

@azrikahar Unfortunately I get the same error as before. Directus does now show my field as type JSON in the UI though.

**Comment by azrikahar** [comment]
Thanks to [this helpful SO answer](https://stackoverflow.com/a/54047683), seems like the returning of the object-like value is done by the [postgres-interval](https://github.com/bendrucker/postgres-interval) package. It does conveniently offer helpers like `toPostgres()` and `toISOString()`, but the issue now would be how can we ensure it to use/return these instead of the original object-like syntax.

Following that trail upstream, we would need to add our custom parser to `pg-types` (which is used in pg client's `types` property). Currently there isn't really a straightforward way, at least as far as I understood... but we can use a custom extension to make this happen:

> ⚠ This is by no means an "official" recommendation and should be viewed as an unofficial workaround unless noted otherwise

- Create a hooks extension, eg `/extensions/interval/index.js`

- Use the following code in the newly created `index.js` file:

  ```js
  module.exports = (_, { database }) => {
	  // extra safety to make sure this extension is only applied when using Postgres
	  if (database.client.constructor.name !== 'Client_PG') return;
  
	  const types = database.client.driver.types;
	  const existingIntervalParser = types.getTypeParser(types.builtins.INTERVAL);
	  const newIntervalParser = (val) => {
		  // this is what the original parser does which returns an object, eg. { hours: 2, minutes: 30 }
		  const parsed = existingIntervalParser(val);
  
		  // You can also use this to return it in ISO format, eg. "P0Y0M0DT2H30M0S"
		  // return parsed.toISOString();
  
		  // We return it as an interval string, eg. 30 minutes 2 hours
		  return parsed.toPostgres();
	  };
  
	  // Override the original Interval parser with our modified parser
	  types.setTypeParser(types.builtins.INTERVAL, newIntervalParser);
  };
  ```

The above steps does seem to ensure it is returned as a string across the board:

- REST

  |Before|After|
  |---|---|
  |![chrome_qHptKqfd24](https://user-images.githubusercontent.com/42867097/162888698-41d1c551-fbc1-4859-9d8d-ce88e623abba.png)|![chrome_pFExg3kUFe](https://user-images.githubusercontent.com/42867097/162888710-7ea932a6-e9ba-4c10-9b70-51af067663b1.png)|

- GraphQL

  |Before|After|
  |---|---|
  |![chrome_FWQXDWPKfO](https://user-images.githubusercontent.com/42867097/162888773-34e6e8ac-8ef4-4989-9228-c7ff1481e639.png)|![chrome_Yep8bFRtaw](https://user-images.githubusercontent.com/42867097/162888781-c1c9b5d3-030e-41c9-8091-f5a3941d866a.png)|

---

@Joey92 would you mind trying the above workaround and see how it goes?

So now the issue would be how to add configuration support for this 🤔 but I reckon this can be something similar to adding support for SQLite spatialite extension, where we can have a more official way of adding this kind of extension as the "hooks" method used above is rather hacky and not actually a hook so to speak.

**Comment by Joey92** [comment]
@azrikahar thanks for checking for a workaround. For now I'm using the REST endpoint for the that specific call, instead of GraphQL. Keeping the response as an object is helpful because my date time lib can directly use it to manipulate the DateTime object.

One could argue though if the value of the field should be represented as an object instead of a scalar value. But that is not my decision to make.

**Comment by rijkvanzanten** [comment]
> This is _technically_ more of a feature request to properly support Interval types in the platform, but lets first treat it as a bug to make sure it's fallback support is returned consistently across both rest and graphql 🙂

I thiiiiiiink I'm circling back on this statement. We're not officially supporting interval values in any capacity, which means that the REST output is simply whatever the `pg` library gives us through `knex`. The only difference between REST and GraphQL is that the REST implementation is a little leaner in unexpected type signatures, and will return whatever the database spits out. GraphQL on the other hand validates both input and output, and will throw the (expected) error that the type doesn't match what Directus expects (which is correct, as Directus hasn't been built to expect intervals in the first place). Moving this to a feature request discussion!

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
